lib/logstash/util/compress.rb

```diff
@@ -0,0 +1,117 @@
+# encoding: utf-8
+require "logstash/namespace"
+require "logstash/environment"
+require "zip"
+require "rubygems/package"
+require "fileutils"
+require "tempfile"
+require "zlib"
+
+module LogStash::Util
+  module Zip
+
+    extend self
+
+    # Extract a zip file into a destination directory.
+    # @param source [String] The location of the file to extract
+    # @param target [String] Where you do want the file to be extracted
+    # Raise IOError If the target directory already exist
+    def extract(source, target)
+      raise IOError.new("Directory #{target} exist") if File.exist?(target)
+      ::Zip::File.open(source) do |zip_file|
+        zip_file.each do |file|
+          path = File.join(target, file.name)
+          FileUtils.mkdir_p(File.dirname(path))
+          zip_file.extract(file, path)
+        end
+      end
+    end
+
+    # Compress a directory into a zip file
+    # @param dir [String] The directory to be compressed
+    # @param target [String] Destination to save the generated file
+    # Raise IOError If the target file already exist
+    def compress(dir, target)
+      raise IOError.new("File #{target} exist") if File.exist?(target)
+      ::Zip::File.open(target, ::Zip::File::CREATE) do |zipfile|
+        Dir.glob("#{dir}/**/*").each do |file|
+          path_in_zip = file.gsub("#{dir}/","")
+          zipfile.add(path_in_zip, file)
+        end
+      end
+    end
+  end
+
+  module Tar
+
+    extend self
+
+    # Extract a tar.gz file into a destination directory.
+    # @param source [String] The location of the file to extract
+    # @param target [String] Where you do want the file to be extracted
+    # Raise IOError If the target directory already exist
+    def extract(file, target)
+      raise IOError.new("Directory #{target} exist") if File.exist?(target)
+
+      FileUtils.mkdir(target)
+      Zlib::GzipReader.open(file) do |gzip_file|
+        Gem::Package::TarReader.new(gzip_file) do |tar_file|
+          tar_file.each do |entry|
+            target_path = File.join(target, entry.full_name)
+
+            if entry.directory?
+              FileUtils.mkdir_p(target_path)
+            else # is a file to be extracted
+              File.open(target_path, "wb") do |f|
+                f.write(entry.read)
+              end
+            end
+          end
+        end
+      end
+    end
+
+    # Compress a directory into a tar.gz file
+    # @param dir [String] The directory to be compressed
+    # @param target [String] Destination to save the generated file
+    # Raise IOError If the target file already exist
+    def compress(dir, target)
+      raise IOError.new("File #{target} exist") if File.exist?(target)
+
+      tar_file = Tempfile.new(File.basename(target))
```

Comment:
Use `Stud::Temporary.file` instead of Tempfile, see https://github.com/jordansissel/ruby-stud/blob/master/lib/stud/temporary.rb#L23 for usage, use the block form so we delete the temporary file when we are done.